Switch on enum type nested in generic type produces LangVersion error

I think this is related to #34678.

**Version Used**: 
master as of April 9 (commit 43f2329f0)

**Steps to Reproduce**:

1. Create a project using non-preview LangVersion
2. Write the following code
```cs
public class C<T>
{
    public enum E
    {
        V1, V2
    }

    public void M()
    {
        switch(default(E))
        {
            case E.V1: // error CS8511
            break;
        }
    }
}
```

**Expected Behavior**:
Builds successfully (as it does in dev15.9)

**Actual Behavior**:
```
CS8511	An expression of type 'C<T>.E' cannot be handled by a pattern of type 'C<T>.E'. Please use language version 'preview' or greater to match an open type with a constant pattern.
```

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs
@@ -210,8 +210,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     {
                         var requiredVersion = MessageID.IDS_FeatureRecursivePatterns.RequiredVersion();
                         if (Compilation.LanguageVersion < requiredVersion &&
-                            // A null pattern can be tested against a type that can be assigned null, even in C# 7.3
-                            !(expression.ConstantValue == ConstantValue.Null && inputType.CanBeAssignedNull()))
+                            !this.Conversions.ClassifyConversionFromExpression(expression, inputType, ref useSiteDiagnostics).IsImplicit)
                         {
                             diagnostics.Add(ErrorCode.ERR_ConstantPatternVsOpenType,
                                 expression.Syntax.Location, inputType, expression.Display, new CSharpRequiredLanguageVersion(requiredVersion));

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs
@@ -2317,5 +2317,31 @@ public class C
                 //         return t is "frog"; // 3
                 Diagnostic(ErrorCode.ERR_ConstantPatternVsOpenType, @"""frog""").WithArguments("T", "string", "preview").WithLocation(17, 21));
         }
+
+        [Fact]
+        [WorkItem(34905, "https://github.com/dotnet/roslyn/issues/34905")]
+        public void ConstantPatternVsUnconstrainedTypeParameter06()
+        {
+            var source =
+@"public class C<T>
+{
+    public enum E
+    {
+        V1, V2
+    }
+
+    public void M()
+    {
+        switch (default(E))
+        {
+            case E.V1:
+                break;
+        }
+    }
+}
+";
+            CreateCompilation(source, options: TestOptions.ReleaseDll).VerifyDiagnostics();
+            CreateCompilation(source, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular7_3).VerifyDiagnostics();
+        }
     }
 }

